cuda/components/prefix_sum.cuh

```diff
@@ -0,0 +1,143 @@
+
+Karlsruhe Institute of Technology
+Universitat Jaume I
+University of Tennessee
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
+ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_CUDA_COMPONENTS_PREFIX_SUM_CUH_
+#define GKO_CUDA_COMPONENTS_PREFIX_SUM_CUH_
+
+
+#include <ginkgo/core/base/std_extensions.hpp>
+
+
+#include "cuda/base/cublas_bindings.hpp"
+#include "cuda/components/cooperative_groups.cuh"
+#include "cuda/components/reduction.cuh"
+#include "cuda/components/thread_ids.cuh"
+
+
+namespace gko {
+namespace kernels {
+namespace cuda {
+
+
+/**
+ * @internal
+ * Calculates the prefix sum of `elements` inside `block_size` sized blocks
+ * in-place. Can be used together with `finalize_prefix_sum` to calculate the
+ * prefix-sum of an array. `block_size` must be the thread block size used when
+ * calling this kernel, which also dictates the size of blocks where the
+ * in-place prefix sum will be calculated `block_sum` is used to store the total
+ * sum of each block. It requires at least `ceildiv(num_elements, block_size)`
+ * elements
+ */
+template <int block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void start_prefix_sum(
+    size_type num_elements, ValueType *__restrict__ elements,
+    ValueType *__restrict__ block_sum)
+{
+    const auto tidx = threadIdx.x + blockDim.x * blockIdx.x;
+    __shared__ size_type prefix_helper[block_size];
+    prefix_helper[threadIdx.x] =
+        (tidx < num_elements) ? elements[tidx] : zero<ValueType>();
+    __syncthreads();
```

Comment:
Since PR #230, we try to prevent the usage of `__syncthreads()` and similar by using [cooperative groups for everything](https://github.com/ginkgo-project/ginkgo/blob/develop/cuda/components/cooperative_groups.cuh).
To do that, you need to `#include "cuda/components/cooperative_groups.cuh"` and do:
```c++
group::this_thread_block().sync();
```

Here, because you use it multiple times, it is better to store the block in a variable and sync with it:
```
auto this_block = group::this_thread_block();
// ...
this_block.sync();
```
And replace `__syncthreads();` with `this_block.sync();`